### ISSUE TITLE
cmd/asm: Go 1.6 asm regression for MOVWQZX table-32768(SP)(R11*2), R15"

### ISSUE THREAD

**Issue by nigeltao** [golang/go#15426]
This line of .s code:

MOVWQZX table-32768(SP)(R11*2), R15

assembles fine with Go 1.4, Go 1.5 and with Go tip, but fails on Go 1.6.2:

```
$ GOROOT=/home/nt/go1.6 ~/go1.6/bin/go test
# github.com/golang/snappy
asm: invalid instruction: 00228 (/home/nt/src/github.com/golang/snappy/encode_amd64.s:343)	MOVWQZX	table+120(SP)(R11*2), R15
asm: asm: assembly of ./encode_amd64.s failed
FAIL	github.com/golang/snappy [build failed]
```

To repeat, it's fixed on Go tip, but one lead might be that the offset changed from -32768 to +120. Perhaps the assembler was confused by SP being both a physical and virtual register.

Given that it assembled properly on Go 1.4 and Go 1.5, should this regression be fixed for Go 1.6.3?

Upstream issue: https://github.com/golang/snappy/issues/29

Upstream workaround: https://github.com/golang/snappy/commit/ec642410cd033af63620b66a91ccbd3c69c2c59a

@randall77 @robpike

**Comment by robpike** [comment]
The assembler is printing that but the diagnostic is actually generated by the obj library and is in new code.

It's line 4188 of https://github.com/golang/go/blob/master/src/cmd/internal/obj/x86/asm6.go and was added by https://go-review.googlesource.com/c/18845/

Since obj is essentially a black box to cmd/asm, it makes sense to start there. Blame says it's @rsc so going there first.

I do not believe I have touched the assembler since 1.6, other than reviewing code, but the infrastructure around it has changed profoundly.

**Comment by robpike** [comment]
The question remains: what fixed it? I haven't figured that out yet.

**Comment by albhaf** [comment]
Did some digging in the history of that file and my unqualified guess is that it was fixed by https://github.com/golang/go/commit/eb0cff9a768373d97dd9b43f8022db9fd61d3b6c (merge from master to dev.ssa)

**Comment by quentinmit** [comment]
Why is this in the Go1.7 milestone if it's already fixed in 1.7 and we're discussing whether it should be in 1.6.3?

**Comment by rsc** [comment]
I wanted to verify that it really was the correct bytes being emitted. I talked to Nigel and he says he's checked and it is correct, so we can leave it at that.
